### PR TITLE
[WIP] Change ZIP file extraction to preserve time stamps

### LINKED ISSUE #420: NuGet do not honor time-stamps when unpacking nupkg files
NuGet uses the current time when unpacking nupkg files, instead of time-stamps stored in the ZIP files.

This leads to "dirty file" problems for build & deployment systems relying on time-stamps for figuring out which header/lib/dll files have changed. An example of this is robocopy-based SW deployment that only replaces new/modified files.

Steps to reproduce:
* With NuGet command-line: "NuGet.exe install zlib" (package found on nuget.org), or
* Create a Visual Studio 2013 C++ project. Add NuGet dependency to "zlib" and build project.
* Verify that unpacked *.h, *.lib & *.dll files are time-stamped with current time instead of the time in the downloaded nupkg files.

### PR DESCRIPTION
**Pull request by forderud**
Work in progress, seeking feedback. Trying to resolve https://github.com/NuGet/Home/issues/420 .

Build fails due to missing "System.IO.Compression.FileSystem" assembly reference(s). Could someone please tell me how to add these, and if the references should be of type ".NET framework" or "DNX Core"?

### CODE REVIEW

**Comment by yishaigalatzer** [comment]
Thanks for jumping on it! We are all going on thanksgiving holiday, so if you don't hear back by eod Monday please reping

File: src/NuGet.Core/NuGet.Common/ZipArchiveExtensions.cs [hunk 1]
```diff
@@ -88,13 +88,7 @@ public static void ExtractFiles(this ZipArchive archive, string targetPath, Func
...
                     }
 
-                    using (var entryStream = entry.Open())
-                    {
-                        using (var targetStream = new FileStream(targetFile, FileMode.Create, FileAccess.Write, FileShare.None))
-                        {
-                            entryStream.CopyTo(targetStream);
-                        }
-                    }
+                    ZipFileExtensions.ExtractToFile(entry, targetFile, /*overwrite*/false);
```
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.Common/ZipArchiveExtensions.cs]: use overwrite: false instead

**Comment by yishaigalatzer** [comment]
> Build fails due to missing "System.IO.Compression.FileSystem" assembly reference(s). Could someone please tell me how to add these, and if the references should be of type ".NET framework" or "DNX Core"?

I'll give it a try and see what's wrong, the references should work for both, these projects are cross compiled.

Ok I see that there is no system.IO.Compression.FileSystem package available on nuget.org. I'll try to figure out where the types go to.

So here is a clue - you can use http://packagesearch.azurewebsites.net/ to find what packages types go to. It looks like the package name is `System.IO.Compression.ZipFile`

File: src/NuGet.Core/NuGet.Common/ZipArchiveExtensions.cs (hunk 1 above)
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.Common/ZipArchiveExtensions.cs]: I believe overwrite should be true (for the case where packages have partially restored and we want to overwrite them)

Comment (forderud) [inline src/NuGet.Core/NuGet.Common/ZipArchiveExtensions.cs]: I've now changed overwrite from false to true in all four functions.

**Comment by forderud** [comment]
MSDN states that I must add a reference to the `System.IO.Compression.FileSystem` assembly (https://msdn.microsoft.com/en-us/library/system.io.compression.zipfileextensions.aspx). I've tried to do this, but I'm unable to figure out which project/json that needs to be updated.

**Comment by yishaigalatzer** [comment]
This is a interesting case, because you are editing a file in a shared package (that provides shared files to other projects), hence the references need to be added to every project that consumes the common package

**Comment by yishaigalatzer** [comment]
@alpaix @forderud - We can take over from here is that is ok with @forderud?

**Comment by forderud** [comment]
That would be great. Thank you @yishaigalatzer.

Handling assembly references proved itself more difficult for me than anticipated, given my lack of experience with DNX and json-based project configuration.

**Comment by yishaigalatzer** [comment]
Closing as @alpaix took over

**Pull request closed without merging**